src/FuncSharp/Extensions/IEnumerableExtensions.tt

```diff
@@ -109,7 +109,24 @@ namespace FuncSharp
             this IEnumerable<<#= CoproductType(i) #>> source,
 <#= Lines(i, x => Indent(12) + "Action<IEnumerable<" + Type(x) + ">> f" + x, separator: ",") #>)
         {
-<#= Lines(i, x => Indent(12) + "f" + x + "(source.Select(c => c." + GetOrdinal(x) + ").Flatten().ToList());") #>
+            var evaluatedSource = source.ToList();
```

Comment:
@KaliCZ not sure we should be doing this here. The original PartitionMatch is ineffecient due to multiple iterations, but this is not much efficient either. IF you want to avoid multiple iterations, you should pass something into this method that's strictly evaluated, it's not a common thing in C# for methods to guarantee single pass. Anyway, to resolve all, I propose to:

- create i lists and i variables at the beginning of the method
- do a single iteration over the collection and categorize the elements into collections
- invoke method for each collection